rskj-core/src/main/java/co/rsk/mine/ForkDetectionDataCalculator.java

```diff
@@ -0,0 +1,117 @@
+import org.ethereum.core.BlockHeader;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ForkDetectionDataCalculator {
+
+    private static final int CPV_SIZE = 7;
+
+    private static final int CPV_JUMP_FACTOR = 64;
+
+    private static final int NUMBER_OF_UNCLES = 32;
+
+    // + 1 because genesis block can't be used since it does not contain a valid BTC header
+    private static final int MIN_MAINCHAIN_SIZE = CPV_SIZE * CPV_JUMP_FACTOR + 1;
+
+    public byte[] calculate(List<Block> mainchainBlocks) {
+        if (mainchainBlocks.size() < MIN_MAINCHAIN_SIZE) {
+            return new byte[0];
+        }
+
+        List<BlockHeader> mainchainBlockHeaders = mainchainBlocks
+                .stream()
+                .map(Block::getHeader)
+                .collect(Collectors.toList());
+
+        return calculateWithBlockHeaders(mainchainBlockHeaders);
+    }
+
+    public byte[] calculateWithBlockHeaders(List<BlockHeader> mainchainBlockHeaders) {
+        if (mainchainBlockHeaders.size() < MIN_MAINCHAIN_SIZE) {
+            return new byte[0];
+        }
+
+        byte[] forkDetectionData = new byte[12];
+
+        byte[] commitToParentsVector = buildCommitToParentsVector(mainchainBlockHeaders);
+        System.arraycopy(commitToParentsVector, 0, forkDetectionData, 0, 7);
+
+        short numberOfUncles = getNumberOfUncles(mainchainBlockHeaders);
+        forkDetectionData[7] = ByteBuffer.allocate(2).putShort(numberOfUncles).array()[1];
+
+        byte[] blockBeingMinedHeight = getBlockBeingMinedHeight(mainchainBlockHeaders);
+        System.arraycopy(blockBeingMinedHeight, 0, forkDetectionData, 8, 4);
+
+        return forkDetectionData;
+    }
+
+    private byte[] buildCommitToParentsVector(List<BlockHeader> mainchainBlocks) {
+        NetworkParameters params = RegTestParams.get();
+        new Context(params);
+
+        long bestBlockHeight = mainchainBlocks.get(0).getNumber();
+        long blockBeingMinedHeight = bestBlockHeight + 1;
+        long cpvStartHeight = (isMultipleOf64(blockBeingMinedHeight) ?
+                                (long)Math.floor((blockBeingMinedHeight - 1) / (double)CPV_JUMP_FACTOR) :
+                                (long)Math.floor(blockBeingMinedHeight / (double)CPV_JUMP_FACTOR))
+                                * CPV_JUMP_FACTOR;
+
+        byte[] commitToParentsVector = new byte[7];
+        for(int i = 0; i < CPV_SIZE; i++){
+            long currentCpvElement = bestBlockHeight - cpvStartHeight + i * CPV_JUMP_FACTOR;
+            BlockHeader blockHeader = mainchainBlocks.get((int)currentCpvElement);
+            byte[] bitcoinBlock = blockHeader.getBitcoinMergedMiningHeader();
+
+            byte[] bitcoinBlockHash = params.getDefaultSerializer().makeBlock(bitcoinBlock).getHash().getBytes();
+            byte leastSignificantByte = bitcoinBlockHash[bitcoinBlockHash.length - 1];
```

Comment:
Make sure the LSB of bitcoinBlockHash is not the MSB for PoW target calculations. I remember the transaction hash or the blockhash was reversed, but I don't remember which of the two was.
If it's the MSB, it will be always 0x00 for Bitcoin mainnet.

Reply:
That's a good one, fortunately I did take that into consideration :)
In the [tests](https://github.com/rsksmart/rskj/pull/857/files#diff-e90903819d3632606b72146875bf115dR282) I'm using a valid bitcoin mainnet header and if you debug them you can see that the value is not 0.